File: java/leetcode/DynamicProgramming.java
package leetcode;
// TOPIC: Dynamic Programming LeetCode Problems | javac DynamicProgramming.java && java leetcode.DynamicProgramming
// Docs: https://leetcode.com/tag/dynamic-programming/

public class DynamicProgramming {

    // TODO 1: Climbing Stairs — LC #70
    //   int climbStairs(int n)
    //   - dp[i] = dp[i-1] + dp[i-2]; optimize to O(1) space with two variables

    // TODO 2: Coin Change — LC #322
    //   int coinChange(int[] coins, int amount)
    //   - dp[i] = min coins to make amount i; dp[0]=0, dp[i]=MIN(dp[i-coin]+1)

    // TODO 3: Longest Increasing Subsequence — LC #300
    //   int lengthOfLIS(int[] nums)
    //   - O(n²) DP: dp[i] = max(dp[j]+1) for j < i where nums[j] < nums[i]
    //   - O(n log n) patience sorting with binary search

    // TODO 4: Longest Common Subsequence — LC #1143
    //   int longestCommonSubsequence(String text1, String text2)
    //   - 2D DP: dp[i][j] = LCS of text1[0..i] and text2[0..j]

    // TODO 5: 0/1 Knapsack (classic)
    //   int knapsack(int[] weights, int[] values, int capacity)
    //   - dp[i][w] = max value using first i items with capacity w

    // TODO 6: Decode Ways — LC #91
    //   int numDecodings(String s)
    //   - dp[i] = ways to decode s[0..i-1]; check single digit and double digit

    // TODO 7: Word Break — LC #139
    //   boolean wordBreak(String s, List<String> wordDict)
    //   - dp[i] = true if s[0..i-1] can be segmented; check dp[j] && dict contains s[j..i]

    // TODO 8: House Robber — LC #198
    //   int rob(int[] nums)
    //   - dp[i] = max(dp[i-1], dp[i-2] + nums[i]); can't rob adjacent houses

    public static void main(String[] args) {}
}

File: ruby/leetcode/dynamic_programming.rb
# frozen_string_literal: true
# TOPIC: Dynamic Programming LeetCode Problems | ruby dynamic_programming.rb

# TODO 1: Climbing Stairs — LC #70
#   def climb_stairs(n)
#     # dp[i] = dp[i-1] + dp[i-2]; optimize to two variables a, b
#   end

# TODO 2: Coin Change — LC #322
#   def coin_change(coins, amount)
#     # dp = Array.new(amount+1, Float::INFINITY); dp[0]=0
#     # dp[i] = coins.reduce(dp[i]) { |min, c| i >= c ? [min, dp[i-c]+1].min : min }
#   end

# TODO 3: Longest Increasing Subsequence — LC #300
#   def length_of_lis(nums)
#     # dp[i] = length of LIS ending at i; O(n²) or O(n log n) with binary search
#   end

# TODO 4: Longest Common Subsequence — LC #1143
#   def longest_common_subsequence(text1, text2)
#     # 2D DP table; dp[i][j] = LCS of text1[0..i-1] and text2[0..j-1]
#   end

# TODO 5: 0/1 Knapsack (classic)
#   def knapsack(weights, values, capacity)
#     # dp[i][w] = max value with first i items and capacity w
#   end

# TODO 6: Decode Ways — LC #91
#   def num_decodings(s)
#     # dp[i] = ways to decode s[0..i-1]; check single and double digit
#   end

# TODO 7: Word Break — LC #139
#   def word_break(s, word_dict)
#     # dp[i] = can s[0..i-1] be segmented; dp[0]=true
#   end
